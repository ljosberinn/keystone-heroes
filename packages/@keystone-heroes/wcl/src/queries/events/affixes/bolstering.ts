import { Affixes } from "@keystone-heroes/db/types";

import type { AllTrackedEventTypes, ApplyBuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const BOLSTERING = 209_859;

/**
 * @see https://www.warcraftlogs.com/reports/mBf1t94GZ2hRWprv#fight=7&type=auras&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applybuff%22%20and%20ability.id%20%3D%20209859%20and%20target.type%20%3D%20%22NPC%22&hostility=1
 *
 * ```gql
 * {
 *   reportData {
 *     report(code: "mBf1t94GZ2hRWprv") {
 *       fights(fightIDs: [7]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5996572, endTime: 7458759, limit: 2000, filterExpression: "type = \"applybuff\" and ability.id = 209859 and target.type = \"NPC\"") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applybuff" and ability.id = ${BOLSTERING} and target.type = "npc"`,
];

const isBolsteringEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: BOLSTERING,
});

type StackDataset = {
  targetID: number;
  targetInstance: number | null;
  stacks: number;
};

const getHighestBolsteringStack = (
  allEvents: ApplyBuffEvent[]
): ApplyBuffEvent[] => {
  const stacksPerNPC = allEvents.reduce<StackDataset[]>(
    (acc, { targetID, targetInstance = null }) => {
      const existingIndex = acc.findIndex(
        (dataset) =>
          dataset.targetInstance === targetInstance &&
          targetID === dataset.targetID
      );

      if (existingIndex > -1) {
        return acc.map((dataset, index) =>
          index === existingIndex
            ? { ...dataset, stacks: dataset.stacks + 1 }
            : dataset
        );
      }

      return [
        ...acc,
        {
          targetID,
          targetInstance,
          stacks: 1,
        },
      ];
    },
    []
  );

  const highestStack = stacksPerNPC.reduce((acc, dataset) =>
    acc?.stacks >= dataset.stacks ? acc : dataset
  );

  const NPCsWithSameFinalStackAmount = stacksPerNPC.filter(
    (dataset) => dataset.stacks === highestStack.stacks
  );

  return allEvents.reduce<(ApplyBuffEvent & { stacks: number })[]>(
    (acc, event) => {
      const dataset = NPCsWithSameFinalStackAmount.find(
        (npc) =>
          npc.targetID === event.targetID &&
          npc.targetInstance === event.targetInstance
      );

      if (!dataset) {
        return acc;
      }

      const previousMatch = acc.find(
        (oldEvent) =>
          oldEvent.targetID === event.targetID &&
          oldEvent.targetInstance === event.targetInstance
      );

      const enhancedEvent = { ...event, stacks: dataset.stacks };

      if (!previousMatch) {
        return [...acc, enhancedEvent];
      }

      // implicitly means it must be a higher stack amount as targetID and
      // targetInstance match
      const isNewer = event.timestamp > previousMatch.timestamp;

      if (isNewer) {
        return [
          ...acc.filter((oldEvent) => oldEvent !== previousMatch),
          enhancedEvent,
        ];
      }

      return acc;
    },
    []
  );
};

export const getBolsteringEvents = (
  allEvents: AllTrackedEventTypes,
  affixSet: Set<Affixes>
): ApplyBuffEvent[] => {
  if (!affixSet.has(Affixes.Bolstering)) {
    return [];
  }

  return getHighestBolsteringStack(allEvents.filter(isBolsteringEvent));
};
