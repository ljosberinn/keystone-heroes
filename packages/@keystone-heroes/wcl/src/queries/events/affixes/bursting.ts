import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const BURSTING = 243_237;

export const getBurstingDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: BURSTING,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
