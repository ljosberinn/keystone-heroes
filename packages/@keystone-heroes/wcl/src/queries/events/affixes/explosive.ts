import type { AnyEvent, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const EXPLOSIVE = {
  unit: 120_651,
  ability: 240_446,
};

/**
 * @see https://www.warcraftlogs.com/reports/YrA4zyZgQJvHbn32#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24target.id%20%3D120651%20AND%20type%20%3D%20%22damage%22%20AND%20overkill%20%3E%200%5E2%24Off%24%23909049%24expression%24target.type%20%3D%20%22player%22%20and%20rawDamage%20%3E%200%20and%20ability.id%20%3D%20240446
 * ```gql
 * {
 *   reportData {
 *     report(code: "YrA4zyZgQJvHbn32") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 87034, endTime: 1527765, limit: 2000, filterExpression: "(target.type = \"player\" and rawDamage > 0 and ability.name = \"Explosion\") OR (target.name = \"Explosives\" AND type = \"damage\" AND overkill > 0)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `target.type = "player" and rawDamage > 0 and ability.id = ${EXPLOSIVE.ability}`,
  `target.id = ${EXPLOSIVE.unit} AND type = "damage" AND overkill > 0`,
];

export const isExplosiveDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: EXPLOSIVE.ability,
});

export const createIsExplosiveDeathEvent =
  (targetID: number) =>
  (event: AnyEvent): event is DamageEvent => {
    return (
      event.type === "damage" &&
      "targetInstance" in event &&
      event.sourceID !== -1 &&
      event.targetID === targetID
    );
  };

export const findExplosiveTargetID = (allEvents: AnyEvent[]): number | null => {
  const dataset = allEvents.reduce<{
    targetID: null | number;
    targetInstance: null | number;
  }>(
    (acc, event) => {
      if (
        // skip damageEvents
        event.type !== "damage" ||
        // ignore event if target is single instance
        !("targetInstance" in event) ||
        // skip empty instances, just TS here
        event.targetInstance === undefined ||
        // check whether the stored instance is lower than this events
        (acc.targetInstance && event.targetInstance < acc.targetInstance)
      ) {
        return acc;
      }

      return {
        targetID: event.targetID,
        targetInstance: event.targetInstance,
      };
    },
    {
      targetID: null,
      targetInstance: null,
    }
  );

  return dataset.targetID;
};
