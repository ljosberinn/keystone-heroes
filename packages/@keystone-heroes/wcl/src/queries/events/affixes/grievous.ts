import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const GRIEVOUS_WOUND = 240_559;

export const getGrievousDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: GRIEVOUS_WOUND,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
