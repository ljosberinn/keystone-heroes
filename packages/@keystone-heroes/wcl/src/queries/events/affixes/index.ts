import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
