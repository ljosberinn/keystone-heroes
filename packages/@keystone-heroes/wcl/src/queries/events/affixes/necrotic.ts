import type { ApplyDebuffStackEvent, DamageEvent } from "..";
import { createIsSpecificEvent } from "../utils";

export const NECROTIC = 209_858;

// const getHighestNecroticStackAmount = (
//   allEvents: ApplyDebuffStackEvent[]
// ): ApplyDebuffStackEvent | null => {
//   const relevantEvents = allEvents.filter(
//     (event): event is ApplyDebuffStackEvent => event.type === "applydebuffstack"
//   );

//   const highestStack = relevantEvents.reduce(
//     (acc, event) => (acc >= event.stack ? acc : event.stack),
//     0
//   );

//   // dont care whether this specific stack size was reached multiple times
//   // just need to track one
//   return relevantEvents.find((event) => event.stack === highestStack) ?? null;
// };

/**
 * @see https://www.warcraftlogs.com/reports/DmVgxdj3WTYBApQt/#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24ability.id%20%3D%20209858%20and%20type%20%3D%20%22applydebuffstack%22%20and%20stack%20%3E%2010%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20209858%20and%20rawDamage%20%3E%200
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DmVgxdj3WTYBApQt") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 0, endTime: 1414930, filterExpression: "(target.type = \"player\" and ability.id = 209858 AND ((type = \"applydebuffstack\" and stack > 10) OR (type = \"damage\" and rawDamage > 0))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
const expressions = {
  base: `target.type = "player" and ability.id = ${NECROTIC}`,
  stack: 'type = "applydebuffstack" and stack > 10',
  damage: 'type = "damage" and rawDamage > 0',
};

export const filterExpression = `(${expressions.base}) AND ((${expressions.stack}) OR (${expressions.damage}))`;

export const isNecroticDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NECROTIC,
});

export const isNecroticStackEvent =
  createIsSpecificEvent<ApplyDebuffStackEvent>({
    type: "applydebuffstack",
    abilityGameID: NECROTIC,
  });

// type NecroticEvents = {
//   damageTakenByNecroticEvents: ReturnType<typeof reduceEventsByPlayer>;
//   highestNecroticStackEvent: ApplyDebuffStackEvent | null;
// };

// export const getNecroticEvents = async (
//   params: GetEventBaseParams
// ): Promise<NecroticEvents> => {
//   const allEvents = await getEvents<DamageEvent | ApplyDebuffStackEvent>({
//     ...params,
//     filterExpression,
//   });

//   const damageTakenByNecroticEvents = reduceEventsByPlayer(
//     allEvents.filter((event): event is DamageEvent => event.type === "damage"),
//     "targetID"
//   );

//   const highestNecroticStackEvent = getHighestNecroticStackAmount(
//     allEvents.filter(
//       (event): event is ApplyDebuffStackEvent =>
//         event.type === "applydebuffstack"
//     )
//   );

//   return {
//     damageTakenByNecroticEvents,
//     highestNecroticStackEvent,
//   };
// };
