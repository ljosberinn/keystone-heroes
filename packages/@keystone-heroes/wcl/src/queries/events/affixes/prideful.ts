import { getCachedSdk } from "../../../client";
import { EventDataType, HostilityType } from "../../../types";
import { getEnemyNPCIDs } from "../../report";
import type { DeathEvent, DamageEvent } from "../types";
import type { GetEventBaseParams, GetSourceIDParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const PRIDE = {
  unit: 173_729,
  aoe: 342_332,
} as const;

export const getManifestationOfPrideSourceID = async (
  params: GetSourceIDParams
): Promise<number | null> => {
  const response = await getEnemyNPCIDs(
    {
      fightIDs: [params.fightID],
      reportID: params.reportID,
    },
    [PRIDE.unit]
  );

  return response[PRIDE.unit] ?? null;
};

export const getManifestationOfPrideDeathEvents = (
  params: GetEventBaseParams<{
    fightID: number;
    sourceID: number;
  }>
): Promise<DeathEvent[]> =>
  getEvents<DeathEvent>({
    ...params,
    dataType: EventDataType.Deaths,
    hostilityType: HostilityType.Enemies,
  });

type ManifestationOfPrideDamageEventsParams<Type extends EventDataType> =
  GetEventBaseParams<{
    targetID: number;
    targetInstance?: number;
    dataType: Type;
  }>;

const getManifestationOfPrideDamageEvents = async <Type extends EventDataType>(
  params: ManifestationOfPrideDamageEventsParams<Type>,
  firstEventOnly = false,
  previousEvents: DamageEvent[] = []
): Promise<DamageEvent[]> => {
  const sdk = await getCachedSdk();
  const response = await sdk.EventData({
    ...params,
    hostilityType: HostilityType.Friendlies,
    limit: 5000,
  });

  const { data = [], nextPageTimestamp = null } =
    response?.reportData?.report?.events ?? {};

  if (data.length === 0) {
    return [];
  }

  if (firstEventOnly) {
    return [data[0]];
  }

  const allEvents: DamageEvent[] = [...previousEvents, ...data];

  if (nextPageTimestamp) {
    return getManifestationOfPrideDamageEvents(
      { ...params, startTime: nextPageTimestamp },
      false,
      allEvents
    );
  }

  return allEvents;
};

type ManifestationOfPrideDamageDoneEventParams = Omit<
  ManifestationOfPrideDamageEventsParams<EventDataType.DamageDone>,
  "dataType"
>;

export const getDamageDoneToManifestationOfPrideEvents = async (
  params: ManifestationOfPrideDamageDoneEventParams,
  firstEventOnly = false
): Promise<DamageEvent[]> => {
  const events = await getManifestationOfPrideDamageEvents(
    {
      ...params,
      dataType: EventDataType.DamageDone,
    },
    firstEventOnly
  );

  if (firstEventOnly) {
    return events;
  }

  return reduceEventsByPlayer(events, "sourceID");
};

type ManifestationOfPrideDamageTakenEventParams = Omit<
  ManifestationOfPrideDamageEventsParams<EventDataType.DamageTaken>,
  "dataType"
>;

export const getDamageTakenByManifestationOfPrideEvents = async (
  params: ManifestationOfPrideDamageTakenEventParams,
  firstEventOnly = false
): Promise<DamageEvent[]> => {
  const events = await getManifestationOfPrideDamageEvents(
    {
      ...params,
      dataType: EventDataType.DamageTaken,
    },
    firstEventOnly
  );

  return reduceEventsByPlayer(events, "targetID");
};
