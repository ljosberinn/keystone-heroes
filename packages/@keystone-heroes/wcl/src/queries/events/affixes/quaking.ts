import { Affixes } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  DamageEvent,
  InterruptEvent,
} from "../types";
import { createIsSpecificEvent, reduceEventsByPlayer } from "../utils";

export const QUAKING = 240_448;

/**
 * @see https://www.warcraftlogs.com/reports/nqrLhBXH6AfC4NdM#fight=27&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22interrupt%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.name%20%3D%20%22Quake%22%5E2%24Off%24%23909049%24expression%24type%20%3D%20%22damage%22%20and%20target.type%20%3D%20%22player%22%20and%20ability.name%20%3D%20%22Quake%22
 * ```gql
 * {
 *   reportData {
 *     report(code: "nqrLhBXH6AfC4NdM") {
 *       fights(fightIDs: [27]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 11446505, endTime: 13247987, filterExpression: "target.type = \"player\" and ability.id = 240448 AND type IN (\"interrupt\", \"damage\")") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */

export const filterExpression = [
  `target.type = "player" and ability.id = ${QUAKING} AND type IN ("interrupt", "damage")`,
];

const isQuakingInterruptEvent = createIsSpecificEvent<InterruptEvent>({
  type: "interrupt",
  abilityGameID: QUAKING,
});

const isQuakingDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: QUAKING,
});

export const getQuakingEvents = (
  allEvents: AllTrackedEventTypes,
  affixSet: Set<Affixes>
): (DamageEvent | InterruptEvent)[] => {
  if (!affixSet.has(Affixes.Quaking)) {
    return [];
  }

  const interrupts = allEvents.filter(isQuakingInterruptEvent);

  const damage = reduceEventsByPlayer(
    allEvents.filter(isQuakingDamageEvent),
    "targetID"
  );

  return [...interrupts, ...damage];
};
