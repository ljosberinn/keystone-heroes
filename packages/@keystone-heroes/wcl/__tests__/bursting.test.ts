import { graphql } from "msw";
import { setupServer } from "msw/node";

import { getBurstingDamageTakenEvents } from "../src/queries/events/affixes/bursting";
import burstingDamageTakenEvents from "./fixtures/burstingDamageTakenEvents.json";

describe("bursting", () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("getBurstingDamageTakenEvents", async () => {
    server.use(
      graphql.query("EventData", (_req, res, ctx) => {
        return res(
          ctx.data({
            reportData: {
              report: {
                events: {
                  data: burstingDamageTakenEvents,
                },
              },
            },
          })
        );
      })
    );

    const result = await getBurstingDamageTakenEvents({
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(result).toMatchSnapshot();
  });
});
