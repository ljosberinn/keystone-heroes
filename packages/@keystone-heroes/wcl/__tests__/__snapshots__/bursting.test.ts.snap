// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBurstingEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 243237,
    "absorbed": 1408,
    "amount": 434085,
    "hitType": 3,
    "mitigated": 185,
    "pin": "0",
    "sourceID": -1,
    "targetID": 5,
    "timestamp": 6025978,
    "type": "damage",
    "unmitigatedAmount": 1593,
  },
  Object {
    "abilityGameID": 243237,
    "absorbed": 1387,
    "amount": 496157,
    "hitType": 3,
    "mitigated": 206,
    "pin": "0",
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 72,
    "targetMarker": 2,
    "timestamp": 6025978,
    "type": "damage",
    "unmitigatedAmount": 1593,
  },
  Object {
    "abilityGameID": 243237,
    "amount": 325053,
    "hitType": 1,
    "mitigated": 689,
    "pin": "0",
    "sourceID": -1,
    "targetID": 3,
    "timestamp": 6025978,
    "type": "damage",
    "unmitigatedAmount": 1593,
  },
  Object {
    "abilityGameID": 243237,
    "amount": 526264,
    "hitType": 1,
    "mitigated": 162,
    "pin": "0",
    "sourceID": -1,
    "targetID": 1,
    "timestamp": 6025978,
    "type": "damage",
    "unmitigatedAmount": 1593,
  },
  Object {
    "abilityGameID": 243237,
    "amount": 482183,
    "hitType": 1,
    "mitigated": 214,
    "pin": "0",
    "sourceID": -1,
    "targetID": 71,
    "timestamp": 6025978,
    "type": "damage",
    "unmitigatedAmount": 1593,
  },
]
`;
